src/jvm/main/org/jetbrains/kotlinx/lincheck/TestGenerator.kt

```diff
@@ -0,0 +1,268 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2023 JetBrains s.r.o.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>
+ */
+
+package org.jetbrains.kotlinx.lincheck
+
+import org.jetbrains.kotlinx.lincheck.execution.ExecutionScenario
+
+
+/**
+ * Generates test for this scenario using determined programming language.
+ */
+internal fun generateTest(scenario: ExecutionScenario, testLanguage: SupportedTestLanguage): String {
+    val testGenerator = when (testLanguage) {
+        SupportedTestLanguage.KOTLIN -> KotlinExecutionScenarioTestGenerator
+        SupportedTestLanguage.JAVA -> JavaExecutionScenarioTestGenerator
+    }
+
+    return testGenerator.generateTestCode(scenario)
+}
+
+/**
+ * Determines programming language used to write the running test using StackTrace
+ *
+ * @return determined programming language or `null` if can't determine language
+ */
+internal fun determineTestLanguage(stackTrace: Array<StackTraceElement>): SupportedTestLanguage? {
+    // take index of last because in LinCheckerKt there may be many methods with the same name
+    val lincheckEntryPointCallIndex = stackTrace.indexOfLast { it.className.startsWith(LINCHECK_PACKAGE_NAME) }
+    if (lincheckEntryPointCallIndex == -1) return null // method call not found
+    if (lincheckEntryPointCallIndex == stackTrace.lastIndex) {
+        return SupportedTestLanguage.KOTLIN // entry point is internal call
+    }
+
+    // get the class where lincheck entry point was called
+    val testClassName = stackTrace[lincheckEntryPointCallIndex + 1].className
+    val testClass = Class.forName(testClassName)
+
+    return if (testClass.isAnnotationPresent(Metadata::class.java)) {
+        SupportedTestLanguage.KOTLIN
+    } else {
+        SupportedTestLanguage.JAVA
+    }
+}
+
+
+internal enum class SupportedTestLanguage {
```

Comment:
private?

Reply:
No, because we pass `testLanguage` into `generateAddCustomScenarioBlock` function.